### ISSUE TITLE
Avoid forced downcasting and unwrapping

### ISSUE THREAD

**Issue by nheagy** [wordpress-mobile/swift-style-guide#7]
Avoid using `as!` to force a downcast, or `!` to force unwrapping.
### Examples

Use `as?` to attempt the cast, then deal with the failure case explicitly

``` swift
// Correct
func process(someObject: Any) {
    guard let element = someObject as? Element else {
        // Optional error handling goes here
        return
    }
    process(element)
}

// Wrong
func process(someObject: Any) {
    process(someObject as! Element)
}
```

Or use `throw` when the failure is unrecoverable

``` swift
func processedItems() throws -> Int {
    guard items = theseThings as? [Item] else {
        throw ThingItemsError.ProcessingError
    }
    return process(items)
}
```
### Rationale

Erica Sadun says ["Every time you use an exclamation point in Swift, a kitten dies"](http://ericasadun.com/2015/11/17/a-handful-of-swift-style-rules-swiftlang/) 

Forced casting and unwrapping can make room for unexpected errors, which are often crashes.  They tend to look like this:

![image](https://cloud.githubusercontent.com/assets/517257/17573323/e18d2162-5f16-11e6-9713-2bb66e613e57.png)

Swift was designed to avoid these kinds of errors with tools like optionals and `throws`. It's often more effort to account for the possibilities but that's as intended 😛 
### Enforcing the rule

Peer reviews are likely the only viable method of enforcement.

**Comment by diegoreymendez** [comment]
I tend to agree with this proposal.

There are a few changes I'd make, though:

- I'd try to simplify the first examples to make them easier to follow:

```swift
// Correct
func process(someObject: Any) {
    guard let element = someObject as? Element else {
        return
    }
    process(element)
}

// Wrong
func process(someObject: Any) {
    process(someObject as! Element)
}
```

- Exceptions are for scenarios we want to handle.  For example failure in data coming from external sources (user input, server responses, maybe even database files).  For internal logical errors I really think we should be using `assert`, `precondition` and `fatalError` instead.  The reason is that if we develop our software to accept those scenarios as possible, we become responsible for testing those paths and maintaining them.  This basically means we need to test something that's not reproduceable by external testers.

Check out [my new proposal](https://github.com/wordpress-mobile/swift-style-guide/issues/8) for a better explanation of what I mean.

**Comment by diegoreymendez** [comment]
I'm assigning myself to this issue, but anyone participating as reviewers can add themselves as well.

**Comment by diegoreymendez** [comment]
To avoid discussing error handling in this proposal, the examples could become:

```swift
// Correct
func process(someObject: Any) {
    guard let element = someObject as? Element else {
        // Optional error handling goes here
        return
    }
    process(element)
}

// Wrong
func process(someObject: Any) {
    process(someObject as! Element)
}
```

**Comment by koke** [comment]
> For internal logical errors I really think we should be using assert, precondition and fatalError instead

I very much agree with the proposal, and we should avoid `!` whenever possible. Sometimes you have to use them though, and if that's the case, it's easier to do `!` than `guard ... else { fatalError() }`, and effectively the same.

The main example that comes to mind is `ImmuTable`, where all the `configureCell` methods have to force cast their cell types:

```
struct MediaSizeRow: ImmuTableRow {
    typealias CellType = MediaSizeSliderCell
    func configureCell(cell: UITableViewCell) {
        let cell = cell as! CellType
```

I do know at build time that `cell` will be a `CellType` but, at least when I implemented it, the generics system was not powerful enough to specify that (I'd have to revisit that for Swift 3). Other cases in the past have been when dealing with Obj-C stuff. 

Some use cases that I think are legit:

- IBOutlets are usually `var foo: Type!`
- `UIImage(named:...)!` / `NSData(contentsOfFile:...)!` (for bundled content)
- `managedObjectContext.existingObjectWithID(...) as! Type`
- ` storyboard.instantiateInitialViewController() as! DomainsListViewController`

**Comment by diegoreymendez** [comment]
> I very much agree with the proposal, and we should avoid ! whenever possible. Sometimes you have to use them though, and if that's the case, it's easier to do ! than guard ... else { fatalError() }, and effectively the same.

At least from my experience they're not the same.  This is how they behave:

- Implicit unwrapping:
  - Messages recorded to log do not include file and line info automatically.
  - Debug: always interrupts, message recorded to log
  - Release: always interrupts, no messages recorded to log.
- Manual unwrapping:
  - Messages recorded to log include file and line info automatically.
  - Debug: all cases show a message
  - Release: interrupts according to developer choice, shows message according to developer choice.

Implicit unwrapping example message:
```
fatal error: unexpectedly found nil while unwrapping an Optional value
```

Manual unwrapping example message:
```
fatal error: Failed a cast that should never fail.: file /Users/diego/Desktop/Testing/Testing/ViewController.swift, line 18
```

**Comment by diegoreymendez** [comment]
In short, the reason I think manual unwrapping is good is because it forces developers to think about whether program interruption should happen or not, and whether logging should happen or not.

**Comment by nheagy** [comment]
Great comments @diegoreymendez @koke! I updated the example as you wrote it, Diego.

**Comment by diegoreymendez** [comment]
@nheagy - Can you create a PR for this proposal?  Feel free to send it my way if you want.

**Comment by koke** [comment]
Sorry I'm late but I have some more comments on it 😉 

> they're not the same. This is how they behave

If the main difference is that there is no file/line info printed, I don't think that's much of a problem since any of the cases will cause a crash which contains a backtrace, and you can get that info from there.

> manual unwrapping is good is because it forces developers to think about whether program interruption should happen or not, and whether logging should happen or not

`as!` and `!` should do that as well. If you're not already thinking if they are a good fit, the app will eventually crash. And as I said above, once crashed I don't care too much about the logging part.

I guess what I'm trying to defend here is that it's good to have a rule to avoid force unwrapping, but sometimes it's a good tool tool use. In those few cases, you should be able to justify it if it in a code review.

**Comment by diegoreymendez** [comment]
> If the main difference is that there is no file/line info printed, I don't think that's much of a problem since any of the cases will cause a crash which contains a backtrace, and you can get that info from there.

You're right.

> as! and ! should do that as well. If you're not already thinking if they are a good fit, the app will eventually crash. And as I said above, once crashed I don't care too much about the logging part.
>
> I guess what I'm trying to defend here is that it's good to have a rule to avoid force unwrapping, but sometimes it's a good tool tool use. In those few cases, you should be able to justify it if it in a code review.

I'm not sure I agree because implicit unwrapping is so accessible, it sometimes feels like we use it out of convenience.  But I don't feel too strongly about it.

Do you think your suggestion can be worded in a way that makes sense for the guide?

**Comment by koke** [comment]
> Do you think your suggestion can be worded in a way that makes sense for the guide?

The current text already says "avoid", which I read as "try not to use it unless it's really a good fit". As long as we're on the same page on that, it's fine as it is.

However, if that's the case and we're not going to have it enforced by SwiftLint, I'd move the rule to the "Must read" section
